mcs/class/System/System/IOSelector.cs

```diff
@@ -155,10 +157,376 @@ public void MarkDisposed ()
+
+										backend_updates [idx] = new BackendUpdate (bu) {
+											operations = GetOperationsForJobs (states [u.handle]),
+										};
+
+										break;
+									case BackendUpdateType.Remove:
+										if (exists)
+											throw new InvalidOperationException (String.Format ("Cannot add handle {0} as existing, while it has previously been removed", (int) u.handle));
+
+										backend_updates.Add (new BackendUpdate (u.handle, BackendUpdateType.Add) {
+											operations = GetOperationsForJobs (states [u.handle]),
+											is_new = true,
+										});
+
+										break;
+									default:
+										throw new InvalidOperationException ();
+									}
+								}
+
+								break;
+							}
+							case UpdateType.Remove: {
+								if (states.ContainsKey (u.handle)) {
+									List<IOSelectorJob> jobs = states [u.handle];
+									for (int j = 0; j < jobs.Count; ++j)
+										jobs [j].MarkDisposed ();
+
+									states.Remove (u.handle);
+
+									int idx;
+									if ((idx = backend_updates.FindLastIndex (bu => bu.handle == u.handle)) == -1) {
+										backend_updates.Add (new BackendUpdate (u.handle, BackendUpdateType.Remove));
+									} else {
+										BackendUpdate bu = backend_updates [idx];
+										switch (bu.type) {
+										case BackendUpdateType.Add:
+											backend_updates.RemoveAt (idx);
+											if (!bu.is_new)
+												backend_updates.Add (new BackendUpdate (u.handle, BackendUpdateType.Remove));
+
+											break;
+										case BackendUpdateType.Remove:
+											throw new InvalidOperationException (String.Format ("Cannot remove handle {0} twice", (int) u.handle));
+										default:
+											throw new InvalidOperationException ();
+										}
+									}
+								}
+
+								for (int j = i + 1; j < updates_last; ++j) {
+									if (updates [j].type == UpdateType.Add && updates [j].handle == u.handle) {
+										updates [j].type = UpdateType.Empty;
+										updates [j].handle = IntPtr.Zero;
+										updates [j].job = default (IOSelectorJob);
+									}
+								}
+
+								break;
+							}
+							default:
+								throw new InvalidOperationException ();
+							}
+						}
+
+						if (updates_last > 0) {
+							updates_last = 0;
+							Array.Clear (updates, 0, updates.Length);
+						}
+
+						for (int i = 0; i < backend_updates.Count; ++i) {
+							BackendUpdate bu = backend_updates [i];
+							switch (bu.type) {
+							case BackendUpdateType.Add:
+								BackendAddHandle (backend, bu.handle, bu.operations, bu.is_new);
+								break;
+							case BackendUpdateType.Remove:
+								BackendRemoveHandle (backend, bu.handle);
+								break;
+							default:
+								throw new InvalidOperationException ();
+							}
+						}
+
+						backend_updates.Clear ();
+
+						Monitor.PulseAll (updates);
+					}
+
+					if (backend_events.Length < states.Count)
+						Array.Resize (ref backend_events, states.Count);
```

Comment:
This looks wasteful. We won't have updates for all sockets on every polling cycle.

We should dynamically grown every time BackendPoll find more events that slots to return them.